src/main/java/array/Array.java

```diff
@@ -0,0 +1,49 @@
+package array;
+
+import java.util.Optional;
+import java.util.function.Consumer;
+
+/**
+ * Interface for an array with undefined number of dimensions
+ *
+ * @param <T> - type of values that allow to put into array
+ * @param <C> - coordinates class for navigation through array
+ */
+public interface Array<T, C> {
```

Comment:
I like your generic approach in here. Aren't you considering to introduce `Coordinates` marker interface for `Coordinates1d` and `Coordinates2d`?

Reply:
great!